proxy/src/main/java/com/velocitypowered/proxy/tablist/VelocityTabListLegacy.java

```diff
@@ -0,0 +1,113 @@
+package com.velocitypowered.proxy.tablist;
+
+import com.google.common.collect.ImmutableList;
+import com.velocitypowered.api.proxy.player.TabListEntry;
+import com.velocitypowered.api.util.GameProfile;
+import com.velocitypowered.proxy.connection.MinecraftConnection;
+import com.velocitypowered.proxy.protocol.packet.PlayerListItem;
+import com.velocitypowered.proxy.protocol.packet.PlayerListItem.Item;
+import java.util.Collections;
+import java.util.Map;
+import java.util.Optional;
+import java.util.UUID;
+import java.util.concurrent.ConcurrentHashMap;
+import net.kyori.text.Component;
+import net.kyori.text.serializer.legacy.LegacyComponentSerializer;
+import net.kyori.text.serializer.plain.PlainComponentSerializer;
+import org.checkerframework.checker.nullness.qual.Nullable;
+
+public class VelocityTabListLegacy extends VelocityTabList {
+
+  private final Map<String, UUID> nameMapping = new ConcurrentHashMap<>();
+
+  public VelocityTabListLegacy(MinecraftConnection connection) {
+    super(connection);
+  }
+
+  @Override
+  public void setHeaderAndFooter(Component header, Component footer) {
+  }
+
+  @Override
+  public void clearHeaderAndFooter() {
+  }
+
+  @Override
+  public void addEntry(TabListEntry entry) {
+    super.addEntry(entry);
+    nameMapping.put(entry.getProfile().getName(), entry.getProfile().getId());
+  }
+
+  @Override
+  public Optional<TabListEntry> removeEntry(UUID uuid) {
+    Optional<TabListEntry> entry = super.removeEntry(uuid);
+    entry.map(TabListEntry::getProfile).map(GameProfile::getName).ifPresent(nameMapping::remove);
+    return entry;
+  }
+
+  @Override
+  public void clearAll() {
+    for (TabListEntry value : entries.values()) {
+      connection.delayedWrite(new PlayerListItem(PlayerListItem.REMOVE_PLAYER,
+          Collections.singletonList(PlayerListItem.Item.from(value))));
+    }
+    entries.clear();
+  }
+
+  @Override
+  public void processBackendPacket(PlayerListItem packet) {
+
+    Item item = packet.getItems().get(0); // Only one item per packet in 1.7
+    String strippedName = PlainComponentSerializer.INSTANCE
+        .serialize(LegacyComponentSerializer.legacy().deserialize(item.getName()));
+
+    switch (packet.getAction()) {
+      case PlayerListItem.ADD_PLAYER:
+        if (nameMapping.containsKey(strippedName)) { // ADD_PLAYER also used for updating ping
+          VelocityTabListEntry entry = entries.get(nameMapping.get(strippedName));
+          if (entry != null) {
+            entry.setLatency(item.getLatency());
+          }
+        } else {
+          UUID uuid = UUID.randomUUID(); // Use a fake uuid to preserve function of custom entries
+          nameMapping.put(strippedName, uuid);
+          entries.put(uuid, (VelocityTabListEntry) TabListEntry.builder()
+              .tabList(this)
+              .profile(new GameProfile(uuid, strippedName, ImmutableList.of()))
+              .latency(item.getLatency())
```

Comment:
The display name still needs to be applied here, which is the text component returned by the legacy component serializer. It's only necessary to apply this if the stripped name and the original don't match (so contains formatting codes).